hal/TARGET_FLASH_CMSIS_ALGO/flash_common_algo.c

```diff
@@ -82,20 +88,38 @@ static int32_t flash_algo_uninit(flash_t *obj, uint32_t address, uint32_t functi
     return ((flash_algo_jump_t)(((uint32_t)&jump_to_flash_algo) | 1))(&arguments);
 }
 
-
+#if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
+__attribute__((cmse_nonsecure_entry))
+#endif
```

Comment:
Isn't this somehow in CMSIS as one line macro that could say (NON SECURE FUNCTION or similar) ? This is in the entire file. 

is __attribute__ valid for all 3 toolchains?

Reply:
Its not a macro in CMSIS,  but its good idea to have this in macro like below: 
```
#if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
#define SECURE_ENTRY_FNC __attribute__((cmse_nonsecure_entry))
#else
#define SECURE_ENTRY_FNC
#endif
```
Will look neat if code below is changed
```
#if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
__attribute__((cmse_nonsecure_entry))
#endif
int32_t flash_init(flash_t *obj)
{
    flash_set_target_config(obj);
    return 0;
}
```
 to
```
SECURE_ENTRY_FNC int32_t flash_init(flash_t *obj)
{
    flash_set_target_config(obj);
    return 0;
}
````

Reply:
@0xc0170 @deepikabhavnani I add `MBED_NONSECURE_ENTRY` for defining all-toolchain secure gateway functions in https://github.com/ARMmbed/mbed-os/pull/6630/commits/1b7b94e0adbc6df71d504838dc879ea65c3bf493 and https://github.com/ARMmbed/mbed-os/pull/6630/commits/92d937dd469a25f2c22503338fc075e5aa9703d3.

Reply:
> Would split be more readable ?

@0xc0170 To split into multiple files, what's the suggested folder structure?
`mbed-os`/`hal`/`TARGET_FLASH_CMSIS_ALGO`/`TARGET_CORTEX_M4`/`flash_common_algo.c`
`mbed-os`/`hal`/`TARGET_FLASH_CMSIS_ALGO`/`TARGET_CORTEX_M23`/`flash_common_algo.c`
`mbed-os`/`hal`/`TARGET_FLASH_CMSIS_ALGO`/`TARGET_CORTEX_M33`/`flash_common_algo.c`
Or other approach?
@deepikabhavnani

Reply:
@ccli8 Regarding split - from the diff here, it looks like there are quite many ifdef in the implementation. Not sure if flash algo are different for cortex families, but rather depending on this security additions?

My main question is - would it be more clear if this additions are put into own file or would just duplicate it and not help?

Reply:
@0xc0170 It seems inevitable to have multiple `#if` statements in single `flash_common_algo.c`. But for split, I have another idea: create another flash cmsis algorithm folder for Cortex-M23/M33 targets:
<pre>
    "NUMAKER_PFM_M2351": {
        "core": "Cortex-M23-NS",
        "default_toolchain": "GCC_ARM",
        "extra_labels": ["NUVOTON", "M2351", "M2351KIAAEES", <b>"FLASH_CMSIS_ALGO_S"</b>],
        "OUTPUT_EXT": "hex",
</pre>

<pre>
mbed-os/hal/TARGET_FLASH_CMSIS_ALGO
<b>mbed-os/hal/TARGET_FLASH_CMSIS_ALGO_S</b>
</pre>

With that, `TARGET_FLASH_CMSIS_ALGO/flash_common_algo.c` keeps unchanged and `TARGET_FLASH_CMSIS_ALGO_S/flash_common_algo.c` would have few `#if` statements in it.

Reply:
If you can create a commit that we could review? wOuld like to compare _S version to regular one

Reply:
@0xc0170 The idea is implemented in https://github.com/ARMmbed/mbed-os/pull/6630/commits/1c51eee2064e7874c6c0906af9adee3362e1faca.

Reply:
@0xc0170 - We are avoiding directory creation for S/NS as it will be difficult to remove later. With these folders in place, it might give false impression that Mbed OS supports both secure / non-secure world.. Current secure library generation is not fully secure and will change with PSA in place. I would suggest to go back to old version with defines, as it will be easier to change them.
 
@ccli8 Sorry for delayed response here.

Reply:
OK, sounds good

Reply:
@0xc0170 @deepikabhavnani OK. I cancel the commit of creating new flash algorithm folder.